### PR TITLE
Add Psalm integration

### LINKED ISSUE #4: Psalm integration
### Feature Request

|    Q        |   A
|------------ | ------
| QA    | yes

#### Summary

As decided during the [Technical-Steering-Committee Meeting on August 3rd, 2020](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling), Laminas wants to implement [vimeo/psalm](https://github.com/vimeo/psalm) in all packages.

Implementing psalm is quite easy.

#### Required
- [ ] Create a `.psalm.xml.dist` in the project root
- [ ] Copy and paste the contents from this [psalm.xml.dist](https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist)
- [ ] Run `$ composer require vimeo/psalm`
- [ ] Run `$ vendor/bin/psalm --set-baseline=psalm-baseline.xml`
- [ ] Add a composer script `static-analysis` with the command `psalm --shepherd --stats`
- [ ] Add a new line to `script:` in `.travis.yml`: `- if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi`
- [ ] Remove phpstan from the project (`phpstan.neon.dist`, `.travis.yml` entry, `composer.json` `require-dev` and `scripts`)

##### Optional
- [ ] Fix as many psalm errors as possible.

### PR DESCRIPTION
**Pull request by rieschl**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

Adds integration of Psalm

fixes #4

### CODE REVIEW

File: src/HandlerFromPipeSpecFactory.php [hunk 1]
```diff
@@ -44,14 +45,16 @@ public function createFromMiddlewareParam(ContainerInterface $container, $middle
...
+        throw new InvalidMiddlewareException(
+            sprintf(
+                'Middleware can be one of: string container id, %s, %s, %s or %s; %s given',
+                MiddlewareInterface::class,
+                RequestHandlerInterface::class,
+                Closure::class,
+                PipeSpec::class,
+                is_object($middleware) ? get_class($middleware) : gettype($middleware)
+            )
+        );
```
Comment (froschdesign) [inline src/HandlerFromPipeSpecFactory.php]: Are these changes related to the Psalm analysis?

File: test/MiddlewareListenerTest.php [hunk 2]
```diff
@@ -53,14 +53,16 @@ public function createMvcEvent(array $matchedParams, array $services = []): MvcE
...
+        $serviceManager = new ServiceManager(
+            [
+                'factories' => [
+                    'EventManager' => function () {
+                        return new EventManager();
+                    },
+                ],
+                'services'  => $services,
+            ]
+        );
```
Comment (froschdesign) [inline test/MiddlewareListenerTest.php]: The same here.

File: src/HandlerFromPipeSpecFactory.php (hunk 1 above)
Comment (rieschl) [reply src/HandlerFromPipeSpecFactory.php]: yeah, it's the auto-formatting of the IDE. I haven't really reviewed the code myself, yet. I just opened a draft PR so no one else starts to work one the same thing.
I'll mark the PR as ready for review, when it's finished. okay? 🙂 
I think you can also add the PR to the [2.0.0](https://github.com/laminas/laminas-mvc-middleware/milestone/1) milestone. @Xerkus wants to release 2.0 with psalm integration. Thanks!

Comment (froschdesign) [reply src/HandlerFromPipeSpecFactory.php]: @rieschl 
> yeah, it's the auto-formatting of the IDE.

Please revert these changes otherwise it will be displayed again as an error when switching to version 2 of the coding standard. See: https://docs.laminas.dev/laminas-coding-standard/v2/coding-style-guide/#4-7-method-and-function-calls

> I haven't really reviewed the code myself, yet. I just opened a draft PR so no one else starts to work one the same thing.
I'll mark the PR as ready for review, when it's finished. okay?

My comment only referred to these changes, everything else is okay. 😃

Comment (rieschl) [reply src/HandlerFromPipeSpecFactory.php]: thanks for the info! Yes I generally try not to do any style changes on not-style PRs.

**Comment by rieschl** [comment]
Okay, this PR is now up to date with the current `2.0.x` branch, including [PHP 8](https://github.com/laminas/laminas-mvc-middleware/pull/11) and [GitHub Actions](https://github.com/laminas/laminas-mvc-middleware/pull/12).
Who wants to review? @froschdesign @Xerkus ? Thanks!

File: test/Integration/MiddlewareDispatchTest.php [hunk 3]
```diff
@@ -78,11 +83,14 @@ public function testDispatchesMiddleware(): void
...
+        assert($sharedEm !== null);
         $services = $this->application->getServiceManager();
-        $services->get('Request')->setUri('http://example.local/middleware');
+        /** @var Request $request */
+        $request = $services->get('Request');
+        $request->setUri('http://example.local/middleware');
         $services->setService('MiddlewareMock', new Middleware());
 
-        /** @var callable&MockObject $listener */
+        /** @var Closure&MockObject $listener */
```
Comment (Xerkus) [inline test/Integration/MiddlewareDispatchTest.php]: This does not seem right. This is not a Closure.

Comment (Xerkus) [reply test/Integration/MiddlewareDispatchTest.php]: callable object with `__invoke()` should be fine. Psalm bug?

Comment (rieschl) [reply test/Integration/MiddlewareDispatchTest.php]: > 
> 
> callable object with `__invoke()` should be fine. Psalm bug?

Yep. I struggled forever with this one. I either get a `Intersection types must all be objects` error or `callable&MockObject is not a valid type`.
I seems that `callable` can not be used for intersection types. If I use the pipe `|` it complains too (`expects callable, PHPUnit\Framework\MockObject\MockObject&stdClass provided`)

Can I just suppress it? 🙂

Comment (Xerkus) [reply test/Integration/MiddlewareDispatchTest.php]: Can you check psalm issues and report it if not there yet?

Comment (Xerkus) [reply test/Integration/MiddlewareDispatchTest.php]: The way to workaround it "properly" here might be to document listener as callable after the mock object calls were performed

Comment (rieschl) [reply test/Integration/MiddlewareDispatchTest.php]: > The way to workaround it "properly" here might be to document listener as callable after the mock object calls were performed

That works, but then `phpcs` complains `No assignment to $listener variable before or after the documentation comment.`

File: src/MiddlewareController.php [hunk 4]
```diff
@@ -26,6 +30,12 @@
 
 /**
  * @internal
+ *
+ * @property PluginManager         $plugins
+ * @property RequestInterface      $request
+ * @property ResponseInterface     $response
+ * @property EventManagerInterface $event
+ * @property EventInterface        $events
```
Comment (Xerkus) [inline src/MiddlewareController.php]: why those are needed?

File: test/Integration/ApplicationTrait.php [hunk 5]
```diff
@@ -27,9 +30,14 @@ trait ApplicationTrait
...
     protected $failOnErrorEvents = true;
 
+    /**
+     * @param array<string, mixed> $extraConfig
+     */
     protected function setUpApplication(array $extraConfig = []): Application
     {
-        $extraConfig['service_manager']['services'][SendResponseListener::class] = new NoopSendResponseListener();
+        $listenerConfig = ['service_manager' => ['services' => [SendResponseListener::class => new Noop()]]];
+        $extraConfig    = array_merge($extraConfig, $listenerConfig);
```
Comment (Xerkus) [inline test/Integration/ApplicationTrait.php]: This change is breaking. It will replace `service_manager` content entirely while it is only meant to override `SendResponseListener` and nothing else. This have to be a recursive merge with stdlib array utils

I am not sure I agree with psalm complains on this one tbh. May be suppress?

File: src/MiddlewareController.php (hunk 4 above)
Comment (Xerkus) [inline src/MiddlewareController.php]: I don't understand why those are needed but $event and $events types are swapped here. Besides, $event is a `MvcEvent`

File: test/MiddlewareControllerTest.php [hunk 6]
```diff
@@ -77,7 +78,7 @@ public function testWillDispatchARequestAndSetResponseFromAGivenRequestHandler()
     {
         $request = new Request();
         $result  = $this->createMock(ResponseInterface::class);
-        /** @var callable&MockObject $dispatchListener */
+        /** @var Closure&MockObject $dispatchListener */
```
Comment (Xerkus) [inline test/MiddlewareControllerTest.php]: Should this also be converted to match the other listener spy change?

**Review by Xerkus — CHANGES_REQUESTED** [review]

**Review by Xerkus — APPROVED** [review]
